File: c++/2023/solutions/day02-2.cpp
#include <iostream>
#include <fstream>
#include <regex>
#include <unordered_map>

int main() {
	std::ifstream input{ "2023/inputs/day02-input.txt" };
	if (!input.is_open()) std::cout << "file did not open\n";
	
	std::string s;
	std::regex r{"(\\d+) (\\w+)"};
	
	int result = 0;
	int curGame = 1;
	while (std::getline(input, s)) {
		auto start = s.find(":") + 2;
		std::smatch m;
		int red = 0;
		int green = 0;
		int blue = 0;
		while (std::regex_search(s, m, r)) {
			int num = std::stoi(m[1]);
			std::string color = m[2];
			
			if (color == "red" && num > red) {
				red = num;
			} else if (color == "green" && num > green) {
				green = num;
			} else if (color == "blue" && num > blue) {
				blue = num;
			}
			s = m.suffix();
		}
		int power = red * green * blue;
		result += power;
	}
	std::cout << result << '\n';
}

File: haskell/2023/solutions/day02-2.hs
import Data.List as L

data Color = Red | Green | Blue | NoCol deriving Eq

getColor :: String -> Color
getColor x
   | isInfixOf "red" x   = Red
   | isInfixOf "green" x = Green
   | isInfixOf "blue" x  = Blue
   | otherwise           = NoCol

spanUntil :: Eq a => (a -> Bool) -> [a] -> ([a], [a])
spanUntil p x 
   | snd s == [] = s
   | otherwise   = ((fst s) ++ [(head $ snd s)], tail $ snd s)
   where s = span p x

minGame :: [String] -> Color -> Int
minGame x c = helper x c 0
   where helper [] c acc = acc
         helper [_] c acc = acc
         helper all@(x:y:xs) c acc
            | col == c   = helper xs c (max (read x) acc)
            | otherwise  = helper xs c acc
            where col = getColor y

power :: String -> Int
power x = (minGame game Red) * (minGame game Green) * (minGame game Blue)
   where game = drop 2 (words x)

solution :: String -> String
solution x = show $ sum $ map power (lines x)

main = do
       file <- readFile "../inputs/2023/day02-input.txt"
       putStrLn $ solution file